velox/functions/prestosql/StringFunctions.h

```diff
@@ -430,4 +430,70 @@ struct LevenshteinDistanceFunction {
   }
 };
 
+// Note: This function newly allocates output using malloc so it should be
+// free'd at the end.
+template <typename T>
+struct NormalizeFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Map for holding normalization form options
+  static inline std::unordered_map<std::string, utf8proc_int16_t>
+      normalizatonMap{
+          {"NFC", (UTF8PROC_STABLE | UTF8PROC_COMPOSE)},
+          {"NFD", (UTF8PROC_STABLE | UTF8PROC_DECOMPOSE)},
+          {"NFKC", (UTF8PROC_STABLE | UTF8PROC_COMPOSE | UTF8PROC_COMPAT)},
+          {"NFKD", (UTF8PROC_STABLE | UTF8PROC_DECOMPOSE | UTF8PROC_COMPAT)}};
+
+  FOLLY_ALWAYS_INLINE void initialize(
+      const core::QueryConfig& /*config*/,
+      const arg_type<Varchar>* /*string*/,
+      const arg_type<Varchar>* form) {
+    VELOX_USER_CHECK(form != nullptr);
+    VELOX_USER_CHECK(
+        normalizatonMap.find(*form) != normalizatonMap.end(),
+        "Normalization form must be one of [NFD, NFC, NFKD, NFKC]");
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& string) {
+    doCall(result, string, "NFC");
+  }
+
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& string,
+      const arg_type<Varchar>& form) {
+    doCall(result, string, form);
+  }
+
+  FOLLY_ALWAYS_INLINE void doCall(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& string,
+      const arg_type<Varchar>& form) {
+    utf8proc_uint8_t* output = nullptr;
+    auto outputLength = utf8proc_map(
+        (utf8proc_uint8_t*)string.data(),
+        string.size(),
+        &output,
+        normalizatonMap.at(form));
+    if (outputLength < 0) {
```

Comment:
Did you check if Presto java returns these errors as well ? If the intent was to validate outputLength then just VELOX_CHECK_GE(outputLength, 0) is sufficient.

Reply:
On the Java side, it seems that the input string is returned unchanged in the case an invalid UTF-8 string is passed, so no errors thrown.

Reply:
Should I follow the same approach and return the original string unchanged in the case output length < 0 meaning errors occurred while using `utf8proc_map` func ?

Reply:
@pdabre12 : We should follow Java semantics.